src/screens/DiscoveryScreenV2/hooks/useDiscovery.js

```diff
@@ -1,13 +1,92 @@
 import React from 'react';
-import {getFollowing} from '../../../service/profile';
+import {Context} from '../../../context';
+import DiscoveryAction from '../../../context/actions/discoveryAction';
+import following from '../../../context/actions/following';
 import {getFollowedDomain} from '../../../service/domain';
+import {getFollowing} from '../../../service/profile';
 import {getFollowingTopic} from '../../../service/topics';
-import following from '../../../context/actions/following';
-import {Context} from '../../../context';
 
 const useDiscovery = () => {
   const [, followingDispatch] = React.useContext(Context).following;
   const [refreshing, setRefreshing] = React.useState(false);
+  const [discovery, discoveryDispatch] = React.useContext(Context).discovery;
+
+  const users = React.useMemo(() => {
+    return discovery.initialUsers.map((item) => ({
+      ...item,
+      following: item.following !== undefined ? item.following : item.user_id_follower !== null
+    }));
+  }, [discovery]);
+
+  const exhangeFollower = (newUserLists, willFollow, userId) => {
+    const indexUser = newUserLists.findIndex((item) =>
+      item.user ? item.user.user_id === userId : item.user_id === userId
+    );
+    if (newUserLists[indexUser].user) {
+      newUserLists[indexUser].user.following = !!willFollow;
+    } else {
+      newUserLists[indexUser].following = !!willFollow;
+    }
+    return newUserLists[indexUser];
+  };
+
+  const topicExchangeFollower = (newTopicLists, willFollow, topicId, fromTopicPage) => {
+    let indexTopic;
+    if (fromTopicPage) {
+      indexTopic = newTopicLists.findIndex((item) => item.name === topicId);
+    } else {
+      indexTopic = newTopicLists.findIndex((item) => item.topic_id === topicId);
+    }
+    newTopicLists[indexTopic].following = !!willFollow;
+    return newTopicLists[indexTopic];
+  };
+
+  const mapUser = (newUser) => {
+    return discovery.initialUsers.map((user) => {
+      if (user.user) {
+        if (user.user.user_id === newUser.user.user_id) return newUser;
+      } else if (user.user_id === newUser.user_id) return newUser;
+      return user;
+    });
+  };
+
+  const updateFollowDiscoveryContext = (willFollow, item) => {
+    const newFollowedUsers = [...users];
+    const newUser = exhangeFollower(
+      newFollowedUsers,
+      willFollow,
+      item.user ? item.user.user_id : item.user_id
+    );
+
+    DiscoveryAction.setDiscoveryInitialUsers(mapUser(newUser), discoveryDispatch);
+  };
+
+  const topics = React.useMemo(() => {
+    return discovery.initialTopics.map((item) => ({
+      ...item,
+      following: item.following !== undefined ? item.following : item.user_id_follower !== null
+    }));
+  }, [discovery.initialTopics]);
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

Similar to the first comment, the `topics` memoization logic seems to be assuming that if `item.following` is undefined, then the topic is followed if `item.user_id_follower` is not null. This might not always be correct and could lead to incorrect follow status being displayed. It would be better to ensure that the backend always sends a valid `following` status for each topic.

```diff
- following: item.following !== undefined ? item.following : item.user_id_follower !== null
+ following: item.following
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->